app/controllers/api/v1/studies_controller.rb

```diff
@@ -531,159 +531,49 @@ def destroy
         end
 
         # begin determining sync status with study_files and primary or other data
         begin
-          # create a map of file extension to use for creating directory_listings of groups of 10+ files of the same type
-          @file_extension_map = {}
-          workspace_files = ApplicationController.firecloud_client.execute_gcloud_method(:get_workspace_files, 0, @study.bucket_id)
-          # see process_workspace_bucket_files in private methods for more details on syncing
-          process_workspace_bucket_files(workspace_files)
-          while workspace_files.next?
-            workspace_files = workspace_files.next
-            process_workspace_bucket_files(workspace_files)
-          end
+          @unsynced_files = StudySyncService.process_all_remotes(@study)
         rescue => e
-          ErrorTracker.report_exception(e, current_api_user, @study, params.to_unsafe_hash)
-          MetricsService.report_error(e, request, current_api_user, @study)
-          logger.error "Error syncing files in workspace bucket #{@study.firecloud_workspace} due to error: #{e.message}"
-          render json: {error: "Unable to sync with workspace bucket: #{view_context.simple_format(e.message)}"}, status: 500
-        end
-
-        files_to_remove = []
-
-        # before saving unsynced directories, make a pass to check if there were any files that ended up as study_files
-        # that should have been in a directory (might happen due to cutoff in files.next when iterating)
-        @unsynced_files.each do |study_file|
-          file_ext = DirectoryListing.file_extension(study_file.name)
-          directory = DirectoryListing.get_folder_name(study_file.name)
-          # check unsynced directories first, then existing directories
-          existing_dir = (@unsynced_directories + @directories).detect {|dir| dir.name == directory && dir.file_type == file_ext}
-          if !existing_dir.nil?
-            # we have a matching directory, so that means this file should be added to it
-            file_entry = {'name' => study_file.name, 'size' => study_file.upload_file_size, 'generation' => study_file.generation}
-            files_to_remove << study_file.generation
-            unless existing_dir.files.include?(file_entry)
-              existing_dir.files << file_entry
-            end
-          end
+          ErrorTracker.report_exception(e, current_user, @study, params)
+          MetricsService.report_error(e, request, current_user, @study)
+          logger.error "#{Time.zone.now}: error syncing files in workspace bucket #{@study.firecloud_workspace} due to error: #{e.message}"
+          redirect_to merge_default_redirect_params(studies_path, scpbr: params[:scpbr]),
+                      alert: "We were unable to sync with your workspace bucket due to an error: #{view_context.simple_format(e.message)}.  #{SCP_SUPPORT_EMAIL}" and return
```

Comment:
It could be done, but I'm not sure its worth the refactoring, since things like `current_user` and `params` are local variables that would have to get passed, and you'd end up with method calls that would be the same amount of duplicated code in both places.